### ISSUE TITLE
Push-to-talk voice in channel chat

### ISSUE THREAD

**Issue by BloodyIron** [vector-im/riot-web#5993]
Whether it's Freeswitch or Jitsi (AFAIK Freeswitch is on the way out), having Push-To-Talk for channel chat would be seriously appreciated! Right now if you have a channel with a whole lot of people wanting to talk, it can turn into a really big mess.

Also, this PTT feature would need to work when the "app" doesn't have focus. I'm thinking this may not work unless the "desktop" version is used, but I'm not 100% up to speed with browser features like this.

It might also be a good idea to get a keyboard bind to toggle self-muting of your own microphone. Let's say, if your fat cat knocks a glass of water all over your compie, oh noes!

**Comment by t3chguy** [comment]
> Also, this PTT feature would need to work when the "app" doesn't have focus. I'm thinking this may not work unless the "desktop" version is used, but I'm not 100% up to speed with browser features like this.

yes this would not be possible in-browser, Discord has the same issue

**Comment by BloodyIron** [comment]
@t3chguy yeah, I wasn't sure if that particular detail was "solved" by browser devs since I last looked at it. ;)

**Comment by t3chguy** [comment]
as far as I'm aware its done by intent, so applications can't behave like keyloggers

**Comment by BloodyIron** [comment]
@t3chguy that... is actually completely reasonable! I hadn't thought of that. Please don't keylog me Facebook! ;D

**Comment by turt2live** [comment]
ftr this is listed in the voip metabug: https://github.com/vector-im/riot-web/issues/3025#issuecomment-287929855

**Comment by BloodyIron** [comment]
@turt2live when I found that thread, I wasn't sure if the PTT facet would get lost in the size of that thread. :( So, I opened this one.

Along the lines of the reason why we have individual issues, instead of just one big meta issue for every single thing people want. ;P

**Comment by turt2live** [comment]
I'm not declaring a duplicate, just providing a reference so people can find it more easily.

**Comment by BloodyIron** [comment]
@turt2live you monster! :O

**Comment by josephtocci** [comment]
Nextel PTT (Always online, push notifications with audio, Example: Zello)
Remember this? I was too young to have a phone, but my parents each had one and the PTT was far above everything else at the time. There were no smartphones at the time. It was faster than calling, and usually the message got through. These days texting won out, partially because if you put your phone down and came back to it you could just read it. PTT you might miss. The other reason texting won out is because you don't necessarily want the person you are talking too IRL to hear what the person on the phone said. There are pros and cons to this method. There is a decent following on Zello which does this. In fact I know someone who uses Zello with all his camping buddies.

Mumble PTT (Must be running, no push notifications for audio, login/join/accept call to use and hear PTT)
This one is more what people will do now. Especially gamers. Even my friend on Zello would probably use this instead. The way he uses Zello is to schedule an hour with everybody, and everybody goes online at the same time and talks. There is no reason to use the always on Zello if you are just going to chatroom with voice. Basically you just mute the microphone, and have a keyboard button to hold down to unmute. (Headphone button on phone) If someone is trying to be quiet IRL and can't get around it, he can plug his headphones in to hear, and type in the room instead of talking. Optionally a text-to-speech thing would be cool, but that is a separate subject.

My thoughts:
I think the Mumble type PTT is what people actually seem to need. You set up a room for the game, and everyone logs on and plays the game. (Not necessarily games but games is the common example) Text is king unless you are doing something with your hands, if you are doing something with your hands, set up a call. It's not complicated.

So I recommend a keyboard button for desktop and headphone button for phone would be excellent shortcuts to temporarily unmute the microphone while it is held down for both native calls and jitsi conferencing. I would recommend Riot to certain people if it had that.

**Comment by t3chguy** [comment]
> headphone button for phone
/me points at vector-im/riot-ios and vector-im/riot-android

**Comment by josephtocci** [comment]
Android issue created:
https://github.com/vector-im/riot-android/issues/1905

iPhone issue created:
https://github.com/vector-im/riot-ios/issues/1736

**Comment by josephtocci** [comment]
It seems my issue created for Android was both marked low priority and closed without being done, so I guess it won't be happening anyway.
https://github.com/vector-im/riot-android/issues/1905

**TL/DR, turn the camera on when you press the unmute button, unless you want an audio only conference call. Would reduce load**

You could even make the same button to unmute the microphone to turn on the camera. I don't necessarily want my camera on all the time, but when I'm talking I want it on. So have the camera on but don't send anything to matrix unless I hold down the PTT button to unmute the microphone. You could make this the default for conference calls and not for 1-on-1. Does that solve the infrastructure problem? Maybe sometime in the future add the Nextel PTT, it is awesome, but if you say it's too much for now then ok.

It reduces load, and you can have an audio PTT only/no camera conference option for gamers, and a full conference call option for people that have 20 eyes to look at 20 different camera feeds for their friends. lol. Probably the best option considering infrastructure, and nothing has to change server side because load should go down even if you add a healthy amount of users.

**Comment by BloodyIron** [comment]
So any word when we can see push to talk for riot?

**Comment by progserega** [comment]
> Also, this PTT feature would need to work when the "app" doesn't have focus.

May be set "autoplay" options for incoming ptt message as options and by default is off? Because if you in bus and read conference in matrix and at this time anybody send ptt-record to conference - all people in bus - will hear this message.

Otherwise, if you go in car/bike and your friends also go in his cars/bikes and you create conference for hear your frends ptt-mesage - at this situation riot must play ptt-message always, even if app have not focus and display is off.

Result: for confidential ptt autoplay must be off  by default in room, but may be set at six variants:
1. No autoplay ptt-message (default)
2. Autoplay ptt message when riot app is foreground (app have focus)
3. Always autoplay ptt message (and when display is off) - by 1 hour
4. Always autoplay ptt message (and when display is off) - by 2 hour
5. Always autoplay ptt message (and when display is off) - by 4 hour
6. Always autoplay ptt message (and when display is off) - forever (use with caution)

**Comment by BloodyIron** [comment]
Where is this on the radar? I know it's listed p3, but I'm not sure how that falls in terms of timeline of being looked at.

**Comment by t3chguy** [comment]
This would have to be implemented by Jitsi.

**Comment by BloodyIron** [comment]
Well, hoping I can appeal to them : https://github.com/jitsi/jitsi-meet/issues/3464

If anyone wants to help, please post her and in the linked jitsi thread. This really _needs_ to get implemented. Not being able to change the shortcut is pretty unacceptable.

**Comment by anoadragon453** [comment]
Just a heads up that I'm making major progress with this, and hope to have basic PTT functionality implemented very soon :)

**Comment by anoadragon453** [comment]
Got a very good chunk done tonight. Getting the functionality working required some extra PRs in other projects, but now I have everything I need to see it through.

Just working out the last fiddly bits, but I will do so tomorrow :)

**Comment by anoadragon453** [comment]
So I've managed to get proper functionality working! You can see a short and silly demo here: https://matrix.to/#/!DdJkzRliezrwpNebLk:matrix.org/$1541185902690rkqXW:amorgan.xyz

There's currently two problems however:

1. You're currently unable to edit the shortcut
2. The shortcut is registered when you add a Jitsi widget, not when you join in to the call, meaning it might not register in all cases

I plan to fix both of those (hopefully today) by integrating into Riot's settings, which I should be able to use to just register the shortcut on start-up and/or when the functionality is enabled.

Accomplishing all this required some PRs in an extra repo, notably https://github.com/WilixLead/iohook, and as such we're currently waiting for those to make it to master, which one already has.

So yes, functionality is nearing close, exciting!

**Comment by BloodyIron** [comment]
@anoadragon453 wait... you can link to Riot URLs, didn't even know that...

As for your work, woot! Awesome!

Does this work when the app is minimized/not in focus?

Also, for the shortcut change, are you going to push that upstream to jitsi as per your comment in : https://github.com/jitsi/jitsi-meet/issues/3464 ?

Woot! Thanks for your hard work @anoadragon453 ! :DD

**Comment by anoadragon453** [comment]
Thank you!

> Does this work when the app is minimized/not in focus?

Yep!

> Also, for the shortcut change, are you going to push that upstream to jitsi as per your comment in : jitsi/jitsi-meet#3464 ?

So turns out the jitsi people were right and we can just use their iframe API to trigger the change. They do, unhelpfully, only have a function for toggling audio, so for muting we first have to query whether the audio is already muted, and if not then toggle.

So essentially what I'm doing here is useful for electron apps that want to embed Jitsi and have PTT functionality to control it, but I'm not actually changing anything in the jitsi project itself.

**Comment by BloodyIron** [comment]
@anoadragon453 okay but I was also asking about the changing of the PTT key binding, my understanding is that needs to be added upstream to jitsi, is that not the case?

**Comment by anoadragon453** [comment]
Ahh, well since we're controlling it from Riot the keybinding actually happens in Riot. We aren't sending a "space" to Jitsi, we're sending a command to toggle mic audio from their API, so making the existing keybinding changeable in jitsi was not something that was necessary to accomplish it in Riot.

However, while jitsi people have said their shortcut code isn't very extensible, I assume someone could probably hack in a new item into the menu to change the shortcut.

Not promising anything on that end, but I am learning a lot about how a process like that may go with this project, so perhaps one day in the future?

In the meantime, use Jitsi through Riot and everything will be wonderful :D

**Comment by BloodyIron** [comment]
@anoadragon453 I primarily care about riot, I just didn't realise we could change the keybind from the riot end. I like where this is going! Yay! :D

**Comment by anoadragon453** [comment]
If you want an update, I'm currently getting close to implementing enabling the functionality and changing the shortcut in settings.

How to record what shortcut the user is currently pressing is still a bit of an unknown for me. I might try to make it work with iohook, or perhaps instead use a separate library as the user will have Riot in focus anyways. The only issue with the latter is iohook seems to use a different encoding for keys than most other libraries, so `Shift - A` in one lib may be `Ctrl - 7` in iohook.

Anyways, I'll try stuff out and report back on how it goes!

**Comment by anoadragon453** [comment]
Alright! So after the work of tonight we have a nearly complete spot in Settings for push to talk now. It has a enable/disable toggle, a display of what the current shortcut is, and a button to set that shortcut. One thing that still needs to be fixed on the enabling/disabling front is enabling the shortcut on startup if you've enabled the option before already. Not entirely sure where to do this yet, but I'll likely find a spot soon.

In addition, I still need to figure out a way to record what shortcut keys are currently being pressed and then assign that to activate push-to-talk, but with the way I've set things up it should be as simple as dropping in a library/function into the right spot.

Some additional things that may be useful (but possibly not in scope of this PR) are:

* Setting Jitsi audio to mute-by-default when push-to-talk is on
* Having a separate settings menu where you can customize all the shortcuts in Riot
* Ability to set whether the shortcut should get captured by Riot or continued to pass-through to the OS while still toggling PPT

**Comment by ara4n** [comment]
(one thought: please make sure that the iohook stuff is only enabled if the user has explicitly turned on PTT. and please apply all possible paranoia to minimise the risk of a JS exploit in Riot somehow giving an attacker a route to turning the app into a systemwise keylogger through the magic of iohook...)

**Comment by anoadragon453** [comment]
@ara4n Very good point! Selective enablement can certainly be arranged. I'll need to look in to how far a JS exploit in electron could go. If it would allow for IPC calls then that would not be fun.

**Comment by ara4n** [comment]
i haven't looked into how iohook works at all, but i'm assuming that you're configuring it from JS to say "oh hai, please tell me whenever the user hits alt-space no matter which app is in the foreground".  at which point a JS RCE could presumably say "oh hai, i'm interested in all the other keycodes too". So i'm  wondering the native code of iohook has some way of being hardcoded to prevent that from ever happening (e.g. only ever letting you register a single keycode to intercept).

(This said, a JS RCE in electron is very bad news anyway, given there may be exploits which can escalate into running native code too, like the https://electronjs.org/blog/chromium-rce-vulnerability trainwreck. But we should still try to safeguard a trivial XSS from being able to break outside the web env via native extensions like iohook...)

**Comment by anoadragon453** [comment]
Seeing as I've thrown a few PRs at iohook already, I could easily add one that limits how many shortcuts you can set at once, perhaps in the package.json? Is that modifiable from electron?

**Comment by BloodyIron** [comment]
Oh, also, any chance we can get Mouse button support for PTT? :)

**Comment by anoadragon453** [comment]
We're definitely aiming for it!

**Comment by anoadragon453** [comment]
Alright, so push to talk in Settings is fully functional now. However, while it saves and registers the shortcut, iohook won't actually pick up on it because it listens for `keycode` instead of `rawcode` in keyboard events. I've created [an issue](https://github.com/WilixLead/iohook/issues/121) to fix this and if not given much fanfare will create a PR for it soon as well. A release for iohook still hasn't gone out since my other PRs, so hopefully this one will make it in before a new release does go out.

But yes, with that we will have keyboard shortcut recording (already done) and then the ability to listen for the recorded shortcut! Unfortunately not that many key combinations are available at the moment, as well as mouse support currently not out, but I'll try to get a PR for all of this up and running and then we can have people with different OS' and setups try things and give feedback.

Finally, this currently only works on [Dimension](https://github.com/turt2live/matrix-dimension), but I'm adding support for Scalar, Riot's default integration manager, as soon as everything else is working. I've already got the dev env set up :)

**Comment by skylord123** [comment]
@anoadragon453 I love you. Keep it up.

**Comment by anoadragon453** [comment]
@skylord123 Thanks!

Alright, I think I've finally figured out a nice solution to the keymapping problem. This whole time I've been trying to record items using a library like Mousetrap so that we can have a nice visual of the current keys that are being pressed. The problem with this is that Mousetrap doesn't have a nice display for many items. Nevertheless, as it had a common mapping with iohook (the `rawcode` property), I went ahead and implemented it, and life was pretty good.

However, the fact that we couldn't properly display all combinations of keys being pressed still bugged me. If you pressed Enter, it would be ' ' as being the current shortcut, and while one could do a lot of regex to make this nicer, it's not an elegant solution.

Cue [KeyboardEvent](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code), which is a thing that exists, and actually *generalizes* to all OS' and international keyboards (as best as it can)! While this is great for display, it unfortunately doesn't have any codes that are in common with what iohook is expecting.

Then I realized, well we could just listen with both at the same time? And since this option is device-specific, if the rawcode of, say, Enter ends up being `65383` and `87` on Windows, that's alright as you'd be setting it per-device.

So, the solution is now, when setting a shortcut, use iohook *and* this fancy in-browser event listener to both display and get the OS/keyboard-specific keycodes for any given shortcut, even supporting fancy ones like Ctrl+ K + Enter + Backspace (for some reason), which wasn't possible with the Mousetrap setup.

As for security, this would require a function allowing iohook to listen to *all* keyboard shortcuts. Fortunately, I think we can limit this to working only when Riot is in-focus, cutting it off if the user decides to go out of focus. This is controlled in the main Electron process, and thus shouldn't be override-able as a result of an Electron XSS to the best of my knowledge (but someone please correct me if otherwise).

In the process of making Mousetrap work, I also submitted another PR to [iohook](https://github.com/WilixLead/pulls), but now it looks like that functionality won't be needed after all. Oh well, hope it's useful to someone :)

**Comment by anoadragon453** [comment]
So the combination of KeyboardEvent and iohook was exactly what was needed, and we can now represent pretty much all keyboard combinations now (mouse to likely following in a separate PR for now :)

Still some bugs to work out, as well as getting this working on Scalar, but as a treat here's a video of it in action! The Missing Translation is due to, well, a translation for this new text not existing yet. And yes, visuals to hopefully be improved before the final version :)

https://streamable.com/ekqf5

**Comment by BloodyIron** [comment]
lol mrtestmanman nice! :D @anoadragon453 

Thanks for your help!

**Comment by anoadragon453** [comment]
![image](https://user-images.githubusercontent.com/1342360/48384235-c152d600-e6e9-11e8-8fe3-4d1719b34c8c.png)

So after a short weekend break, Push to Talk is now completely working! I've done the following today:

* Made PTT initialize when a Jitsi widget is shown, and disabled when the widget disappears.

This is all instead of doing so only when the PTT option is enabled, or when Riot starts, which are both unnecessary. Of course, the option will still enable/disable mid-call

* Stopped recording a new shortcut when Riot loses focus.

There is justified concern of the shortcut recording feature, which records all keypresses when setting a shortcut to see which keys you're pressing. This feature is now always disabled when Riot is not in focus, even in the event of an XSS.

* Stop Riot from crashing when a secondary window was opened.

This one was important, obviously. Riot no longer crashes when you try to open a window, such as when sending a file or clicking on the taskbar icon :)

The final step is just to get the setup working with Scalar, which should happen shortly after I get a working Scalar setup going locally (about 80% of the way there, but need a few more pieces). After that will be the PR phase :)

**Comment by BloodyIron** [comment]
@anoadragon453 

How does this impact those that want the option to have the mic wide open too? Is that just the behaviour when PTT is disabled in settings?

**Comment by anoadragon453** [comment]
If PTT is disabled, it will be the same functionality is it is now.

**Comment by anoadragon453** [comment]
Alright! Most of the day was spent on getting Scalar up and running, but now it finally is, hooray!

So will just need to hack on Scalar a little bit tomorrow and everything should be good to go!

**Comment by anoadragon453** [comment]
Ok, got Scalar working really nicely, and PRs are now live!

* https://github.com/vector-im/riot-web/pull/7709
* https://github.com/matrix-org/matrix-react-sdk/pull/2280

In addition, I managed to make the following work while polishing everything up:

* i18n support, so no more 'missing translation' text
* *Any* key (in any combination) can be used on the keyboard, not just a limited subset
* Distinction between similar keys like `Shift` and `RightShift`

Mouse buttons *might* work (testing wanted!), but I'm envisioning we'll need to do some UI work for that (little picture of a mouse, maybe?).

Look forward to seeing these land on the /develop branch!

**Comment by BloodyIron** [comment]
Can we add to the future to-do list the Riot icon have a mute symbol over it when PTT is muted or something? Although, not sure how that would work for multiple room jitsi chat... uhhhh?

**Comment by anoadragon453** [comment]
> Can we add to the future to-do list the Riot icon have a mute symbol over it when PTT is muted or something?

Sure

> 
Also, when we get the audible cue for PTT mute/unmute sound, can we make that two sounds, and user-customizable and/or pre-defined in the Riot config.json? (for those of us who built it ourselves :D )

That might be part of a big redactor which allows to customize all the different sounds of Riot, but I'll see how difficult it is to just add custom sounds for PTT for now.

> Also, thanks a tonne for this! Yay!

yw!

**Comment by anoadragon453** [comment]
Mainlining has been momentarily stuck behind the concern that our build machines currently run MacOS (which can compile electron builds for Linux/MacOS/Windows), but compiling native node modules (of which iohook is) requires actually building on Linx/MacOS/Windows boxes. We can obviously accomplish this via VMs, but the infrastructure needs to be modified first.

In the meantime, as iohook provides pre-compiled binaries, the decision is to just use these until we have our own systems set up. The binaries are downloaded automatically during the typically `npm install` phase, so the actual build process will not change at all, unless you want to compile iohook on your own (instructions for which should be provided soon).

Other than that, the issues with the Settings UI have been fixed, and here is the current implementation :)

![image](https://user-images.githubusercontent.com/1342360/48837788-95d58880-ed86-11e8-9288-06cf505bb666.png)

**Comment by BloodyIron** [comment]
@anoadragon453 any particular advantage to having iohook be part of the build flow vs binaries?

**Comment by jryans** [comment]
> any particular advantage to having iohook be part of the build flow vs binaries?

I would assume the main advantages are around trust and future maintenance. When you rely on pre-compiled dependencies, you are trusting the dependency author to build them without anything malicious added in. On the maintenance side, they may also stop providing them for new Node / Electron versions. So in the long run, it's best to build from source, especially for security focused software like Riot.

**Comment by anoadragon453** [comment]
Pretty much this.

**Comment by BloodyIron** [comment]
right! any chance iohook will get baked into the main repo here then? (for those of us looking to do our own builds of riot? ;P )

**Comment by anoadragon453** [comment]
Shouldn't need to, the process is just running `npm install` like you normally do, but then going into node_modules/iohook and running `npm run build`.

**Comment by anoadragon453** [comment]
Scalar has just released a new build which includes the PTT patches, which means one no longer needs to run their own integrations manager to get this all working, they can just use the default in Riot.

Should make testing the above PRs much easier for everyone involved.

**Comment by BloodyIron** [comment]
@anoadragon453 I'm probably going to forget that in the future, would that be documented anywhere? So npm install pulls the iohook source too? And the build second step is on the local files?

As for integrations manager, was that necessary for jitsi before or what? I thought that part might still be required?

**Comment by anoadragon453** [comment]
@BloodyIron Haha, we'll definitely provide instructions in our repo, but if you like the build instructions are also mentioned in [iohook's documentation](https://wilix-team.github.io/iohook/manual-build.html).

**Comment by BloodyIron** [comment]
oh btw, did you test PTT with screen sharing at all? does the behaviour change there at all?

So Riot has Jitsi baked in now? (next build release?) so it won't need to load the "widget" when starting/stopping chat, or what? How do you tell it what Jitsi user to use? (self-hosted)

Dimension, the integration server? not quite seeing why you mention it's supported, am I missing something?

**Comment by anoadragon453** [comment]
@BloodyIron It should work regardless of screen sharing mode enabled or not.

> So Riot has Jitsi baked in now? (next build release?) so it won't need to load the "widget" when starting/stopping chat, or what? How do you tell it what Jitsi user to use? (self-hosted)

Mmm, no not really baked in yet, although @Half-Shot has expressed some interest in working on this, and it would be fairly trivial to actually get some of the Riot UI to react to what's happening in the Jitsi call. For the moment the widget is still required though.

> How do you tell it what Jitsi user to use?

You mean Jitsi server? (Username can just be set by double clicking on your square in the top right) For the server, go into Riot's config.json and you can set the Jitsi URL using the `integrations_jitsi_widget_url` key. The default is `https://scalar.vector.im/api/widgets/jitsi.html`.

> Dimension, the integration server? not quite seeing why you mention it's supported, am I missing something?

Some people use Dimension as it includes some more features than Scalar. So for those people, PTT will also work :)

**Comment by BloodyIron** [comment]
Hah yeah I meant to put Jitsi Server, not User. I just wasn't sure if these changes meant Jitsi server declaration changed in some way, as you mentioned [not needing] an integration server by default?

I haven't quite gotten to the point of identifying which integration server I'll use, but I suspect it may be Dimension. Haven't vetted. Good to know this isn't a constraint! :D 

Any idea how soon before this hard work of yours could hit public release? :)

**Comment by anoadragon453** [comment]
So while the PR for this has been out for a while, getting it merged has been blocked on getting a Windows VM up and running and being able to automate builds through it. While something that can likely be done in a few days, the team's time is pretty strapped right now and would rather be spending what time we have on Riot's redesign than on wrangling Window's many awful command line environments.

That said, we have made progress on this this week. We've set up a stripped-down Windows running headless on our build server. We've got an ssh server on it and have it hooked up to our Jenkins infrastructure. At the moment the only blocker has been getting a suitable bash environment for running Riot's setup scripts or just sucking it up and converting them to batch. (We're using git-bash (MINGW32) to execute things, but running stuff in there triggered from a native Windows cmd or Powershell doesn't really work too well, any advice on a less hacky solution would be appreciated).

So that's our current progress - tantalizingly close, and just working on it where we have a spare 30m or so. So if anyone does any advice for getting a proper bash shell that can be ssh'd into and doesn't involve Windows Subsystem Linux, that'd be highly appreciated!

Otherwise we'll probably get it all working in the next few days anyways :) Watch this space!

**Comment by JoshStrobl** [comment]
@anoadragon453 That's fantastic to hear and thanks for the update. Really looking forward to being able to test out PTT :+1:

**Comment by BloodyIron** [comment]
@anoadragon453 cygwin?

**Comment by anoadragon453** [comment]
We ended up just throwing it on a faster machine and installed WSL.

But now we're hitting [this bug](https://github.com/webpack-contrib/uglifyjs-webpack-plugin/issues/272) and builds are timing out :/

**Comment by anoadragon453** [comment]
Build time outs fixed by reducing VirtualBox CPUs to 1 which still complete builds in a reasonable amount of time.

Slowly but surely!

**Comment by BloodyIron** [comment]
@anoadragon453 how on earth does limiting the CPU to 1 solve build timeouts? That's weird man! :O 

But glad to hear progress! :DDD

**Comment by anoadragon453** [comment]
> @anoadragon453 how on earth does limiting the CPU to 1 solve build timeouts? That's weird man! :O

windozzzzz

**Comment by anoadragon453** [comment]
Huzzah! Hot on the heels of [our FOSDEM stuff](https://matrix.org/blog/2019/02/04/matrix-at-fosdem-2019/) we finally got Windows to play nice! We only hit numerous npm bugs (switched to yarn) and [a bug in the ecosystem](https://github.com/vuejs/vue-cli/issues/3407) but all is good now and we have [working Windows builds from Windows](https://matrix.org/jenkins/job/VectorElectronDevelopWin/)!

At this point I just need to update and rebase my PTT PR and then, pending review, you should expect to see it in the client shortly!

**Comment by alien2003** [comment]
Any progress with it? A lot of Discord users don't switch because lack of PTT

**Comment by anoadragon453** [comment]
Hey, thanks for asking. Currently progress is blocked on https://github.com/wilix-team/iohook/issues/151, aka getting the underlying native node module working on Windows with Electron v4. Everything is good to go on the MacOS/Linux side.

I've tried a couple times to get to the bottom of it but to no avail (C++ isn't my main language). If anyone is familiar with C++ and its build chain if they'd like to take a shot at fixing that it would be highly appreciated!

**Comment by BloodyIron** [comment]
Any chance the MacOS/Linux users can get this first? ;D

**Comment by anoadragon453** [comment]
Was thinking about it but last thing we want to do is split our codebase on a platform-by-platform basis :)

**Comment by alien2003** [comment]
Seems like it is fixed

**Comment by anoadragon453** [comment]
@alien2003 Indeed thanks, though I'm getting another bug on our Windows dev box. Going to remake the folder and see how it does.

**Comment by alien2003** [comment]
Hype... Finally we will ditch Discord

**Comment by BloodyIron** [comment]
Pssst any news on this? ;)))

**Comment by anoadragon453** [comment]
Well I spent about 6 hours on this just now and have hit a brick wall on compiling iohook on Windows again. Running `yarn run build` aka `cmake-js` spits out `There is no Visual C++ compiler installed. Install Visual C++ Build Toolset or Visual Studio.` It does seem to be 100% installed though, and the message persists even after reboots.

If anyone can figure out how to build [iohook](https://github.com/wilix-team/iohook) on Windows atm please let me know. I've left build instructions here: https://github.com/anoadragon453/riot-web/blob/anoa/jitsi_ptt/docs/native_node_modules.md#compiling-iohook

Here are my notes for the current session for installing from a fresh Windows install:

	Turn off windows defender and firewall etc
	Install chocolately: get npm, yarn, git, python2, visual C++ Build Tools 2017
	Open Powershell
	Clone iohook
	yarn
        yarn run build
        Error :(

Now, even if we do get it building we'll likely run into this showstopper: https://github.com/wilix-team/iohook/issues/167

I think tomorrow I'll just change the PRs to not load iohook if on Windows because it's been a blocker for far too long and everytime something on Windows is fixed something else comes up again :')

**Comment by BloodyIron** [comment]
@anoadragon453 maybe your PATH value was missing something?

**Comment by ImGunter** [comment]
Can’t wait for this feature to be added thanks for the hard work sir!

**Comment by skylord123** [comment]
@anoadragon453 definitely sounds like a path issue of some sort. 

Check out "Setting up your command-line environment" under this link:
https://devblogs.microsoft.com/cppblog/finding-the-visual-c-compiler-tools-in-visual-studio-2017/

I use windows the most (sadly) and so I am a bit sad to hear about the showstoppers. I'm fine with forgetting about Windows for now as long as it isn't put off for a very long time (of course that isn't entirely in our control if the upstream issues aren't resolved).

Thanks for all the time you have spent on this. I definitely appreciate it :)

**Comment by BloodyIron** [comment]
I think it's in everyone's best interests if we get this feature sorted out for everyone. If not for this feature, at least for streamlining the build process for other things too.

**Comment by W4rH0rse** [comment]
PTT is absolutely critical to alot of gamers who want to move away from Discord, and most of them are on Windows.

**Comment by TaylorEverlund** [comment]
> 
> 
> PTT is absolutely critical to alot of gamers who want to move away from Discord, and most of them are on Windows. I'd love to use Riot as a replacement for Discord, but until PTT is supported I cant.

Ditto, I have friends who've made the move from TS to Discord and now want to move to riot but PTT is holding them back. We've programmed many of mouth breathing food chewers to use PTT and without it would be a nightmare. I wish I could help in some way with the coding on this but with so many other side projects I just don't have the time. I wish y'all the best of luck and for a speedy resolution.

**Comment by ImGunter** [comment]
Any updates on where we are with this feature?

**Comment by anoadragon453** [comment]
It's currently been blocked behind other tasks, but as Matrix 1.0 has just released and all the work that has gone with that, I intend to work on this feature very soon.

**Comment by ImGunter** [comment]
Yay that's great to hear i can't wait for that feature to come out please keep us updated and thank you for your great work!

**Comment by anoadragon453** [comment]
Thanks, I will :)

**Comment by alien2003** [comment]
Not directly related but seems like a good addition to Jitsi PTT feature
https://github.com/Johni0702/mumble-web

**Comment by ImGunter** [comment]
Now that we have reactions and editable messages is push to talk next on the road map?

**Comment by anoadragon453** [comment]
> Not directly related but seems like a good addition to Jitsi PTT feature
> https://github.com/Johni0702/mumble-web

Oh, that already has instructions for being an integration manager widget, interesting!

As long as it has support for [HTML5 cross-document messaging](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) so that we can give it commands from the web interface it should be quite easy to integrate.

**Comment by anoadragon453** [comment]
> Now that we have reactions and editable messages is push to talk next on the road map?

It's getting close :)

**Comment by BloodyIron** [comment]
I'm switching to riot AWAY from Mumble :( Reliable tech, but their devs have dragged their heels (and effectively halted) for well over 7 years now.

**Comment by ImGunter** [comment]
Same bloody mumble is such a great program with great potential but the devs don’t seem to have the time anymore.

**Comment by alien2003** [comment]
> Oh, that already has instructions for being an integration manager widget, interesting!

Yes. I tried it for my Minecraft server room, works well but it's not persistent widget by default. The only way to make it persistent is to set it to "Jitsi" type and it breaks widget in mobile apps

**Comment by BloodyIron** [comment]
Any word on this?

**Comment by anoadragon453** [comment]
Sort of blocked behind not working on Windows. I don't think we want to release a feature that only supports a subset of our supported platforms. Windows may work now though, I haven't checked in a few months :thinking:

**Comment by BloodyIron** [comment]
@anoadragon453 is it the supporting library is limited on Windows, or what? I know this is going to be a FR that won't go away, and I'm not talking about my stubbornness, I'm talking about everyone else's too XD

**Comment by anoadragon453** [comment]
FR?

It's good ol' [iohook](https://github.com/WilixLead/iohook).

**Comment by BloodyIron** [comment]
Feature Request.

**Comment by anoadragon453** [comment]
> Feature Request.

ahh

**Comment by ImGunter** [comment]
Damn iohook is still holding you up that’s annoying smh windows lmao

**Comment by Mhowser** [comment]
Any update to this?

**Comment by anoadragon453** [comment]
> Any update to this?

We still don't have the build infrastructure to produce Windows builds on a Windows box, but I imagine this will change very shortly as we have a number of native node modules coming to Riot Desktop (namely enabling Push-To-Talk and searching in encrypted rooms).

So be on the lookup for that (very hopefully) soon!

**Comment by anoadragon453** [comment]
Some progress! https://github.com/vector-im/riot-web/pull/10369#issuecomment-549865992

We want to get it into the codebase so the PRs stop bitrotting, but people who want to enable it will still need to insert their own native node module in.

That said, things are still moving on upgrading the build infra. Integrating [seshat](https://github.com/matrix-org/seshat) into Riot Desktop for instance has made the priority for native node module support even higher.

**Comment by alien2003** [comment]
@anoadragon453 That's better than nothing but still not enough to make average Discord users move to Riot :( It will be hard to convince Discord users to install Riot if there is no ready-to-use build with PTT stuff included

**Comment by anoadragon453** [comment]
> @anoadragon453 That's better than nothing but still not enough to make average Discord users move to Riot :( It will be hard to convince Discord users to install Riot if there is no ready-to-use build with PTT stuff included

Of course. This is just a stopgap solution to stop the code from bitrotting. Once the build infrastructure is up then all users will have it by default.

**Comment by BloodyIron** [comment]
So, it's coming up to nearly two years for this feature request. And my understanding Windows is the last mile.

Can we _please_ get this finalised and pushed into mainline before year's end? (2019)

**Comment by Mhowser** [comment]
Is there a way we could build a client with this PR to test it?

**Comment by samjco** [comment]
PTT rooms, please!! This would be greatly appreciated

**Comment by samjco** [comment]
Just curious will any of these libraries work?

https://github.com/mnovicio/ptt-freeswitch-ui
https://github.com/Merve40/ptt
https://github.com/QVDev/GunPushToTalk
https://github.com/Theofilos-Chamalis/QR-PTT-PushToTalk

Can the mumble server audio server be tied in somehow and integrated into Riot?
https://www.mumble.info/

So here is my dirty hack.. Install mumble-server.
Hack the mumble web client here: https://github.com/Johni0702/mumble-web
which attaches to mumble. Use iframes with Riot that uses the modified mumber web client.
As a user is added to Riot, auto add/register them to the mumble server. When an audio room is made in Riot, add a room in mumble and point the iframe to it. Yeah, this hack is As dirty as it gets :)

**Comment by samjco** [comment]
Wait! did I just see this towards the bottom!!!

![image](https://user-images.githubusercontent.com/3359062/74011439-74da6280-494d-11ea-8480-f5292698d07d.png)
https://github.com/Theofilos-Chamalis/mumble-web

More notable web themes for mumble:
https://github.com/xPoke/MetroMumble/releases

**Comment by samjco** [comment]
Widget for matrix
https://git.aventer.biz/AVENTER/js-matrix-mumble-web/src/commit/55fb04eb92bc7e934eba99ed381fd714bff62d0e

For iOS:
https://github.com/mumble-voip/mumble-iphoneos

For OSX desktop kit
https://github.com/mumble-voip/mumblekit

Honestly, if Riot has Jitsi integrated, why not just use the video, but cut out the video and make the Push-to-talk button off of the mute button?

**Comment by anoadragon453** [comment]
> Honestly, if Riot has Jitsi integrated, why not just use the video feature, but cut out the video and make the Push-to-talk button off of the mute button?

That's what this PR does :)

But it's bitrotted and I haven't found time to update it :(

But hopefully sometime soon.

**Comment by samjco** [comment]
@anoadragon453 Did you see this?
Like: https://jitsi.org/news/audio-only-mode-in-jitsi-meet/
and here: https://community.jitsi.org/t/audio-only-mode/21359/2

**Comment by anoadragon453** [comment]
@samjco I'm aware of this mode. To be clear, the scope of this PR is just to have a global shortcut to toggle your microphone in a jitsi call, regardless of whether Riot is in focus, as well as tools to allow you to modify that shortcut. There's no audio-only UI here (yet), though this probably would unlock that work.

**Comment by Per0x** [comment]
> **Honestly, **if Riot has Jitsi integrated**, why not just use the video feature, but cut out the video and make the Push-to-talk button off of the mute button?**

I don't recommend jitsi for this type of use. Beyond 3 users, Jitsi doesn't work very well, whereas a voice chat should work well with more than 10 people.

**Comment by samjco** [comment]
@Per0x Jitsi itself has an audio-only-mode:
https://jitsi.org/news/audio-only-mode-in-jitsi-meet/

make a dummy meeting based on unexposed domain and random characters and Just hide the UI. Anyone accessing the room would be auto-logged in.

As for private rooms, the domain + random is exposed.

Just thinking out loud here.

**Comment by BloodyIron** [comment]
What's the max # of users Riot devs have tested in a single Jitsi meeting? What's the upper limit?

**Comment by samjco** [comment]
@BloodyIron depends on the server. some say 10 other 30-40
https://community.jitsi.org/t/jitsi-meet-max-call-connect-on-one-conference/20519

**Comment by BloodyIron** [comment]
I'm more meaning hypothetically where computing/networking resources are not the constraint. At what point would Jitsi crawl to a halt?

**Comment by turt2live** [comment]
> What's the max # of users Riot devs have tested in a single Jitsi meeting? What's the upper limit?

We regularly have ~20-24 people on a single call with no major issues. Earlier this week we got about 20-30 people onto a single call to load test it (though this time the machines trying to send the video had the problem, not the server to my knowledge).

**Comment by BloodyIron** [comment]
> > What's the max # of users Riot devs have tested in a single Jitsi meeting? What's the upper limit?
> 
> We regularly have ~20-24 people on a single call with no major issues. Earlier this week we got about 20-30 people onto a single call to load test it (though this time the machines trying to send the video had the problem, not the server to my knowledge).

How much bandwidth did the 20-30 person call use inbound and outbound? And was that video or audio-only?

**Comment by turt2live** [comment]
It was video and audio. I don't have the server's numbers, but from memory one of the machines responsible for trying to push ~6 streams was hitting about 20mbps upload and about 18mbps download (I know this because it was my machine :p)

**Comment by BloodyIron** [comment]
> It was video and audio. I don't have the server's numbers, but from memory one of the machines responsible for trying to push ~6 streams was hitting about 20mbps upload and about 18mbps download (I know this because it was my machine :p)

6 streams? of webcams or what? How does 20-30 people turn into 6 streams? I'm not quite following you. Trying to plan capacity here.

Sounds like 3mbps per "stream"? (whatever each stream is, I don't yet know from what you are referring here).

**Comment by turt2live** [comment]
The load test wasn't 20-30 physical people. It was 20-30 video streams, mostly webcams. Mine just happens to try and send 1080p video over the internet, but Jitsi will downscale as needed for bandwidth iirc.

If you're planning for load, it's generally going to be best to ask the Jitsi crew specifically and not us. We're a very specific use case and might use Jitsi differently than you (mostly because we're a giant node for everyone who uses the default settings). The server isn't that large though from what I remember.

**Comment by Mathias-g** [comment]
Any news on the push to talk feature? We are thinking about implementing mumble when the MSC for room types gets implemented, and make a "mumble room type", but if push to talk is getting implemented in jitsi meet and the 1 to 1 calls then we won't have to.

**Comment by fti7** [comment]
Mumble Integration would be very nice :-) i think it would be the superior alternative for the "Voice Channels" Feature.
PTT also needs good Voice Activation etc. to be a real alternative. 
This Issue is 2 Years old... dont think they will ever do it ;-)

**Comment by anoadragon453** [comment]
Heh, I was just saying today that I should get back to this. But a lot of the code has bit-rotted, so it'll take some work.

Most of the effort in this feature was getting the native node module for listening to global keystrokes working with the js package for site-scoped keystrokes. That will be useful for jitsi or any other voice activation software.

Tbh, jitsi eats all the resources on my PC so I'm happy to have mumble as a potential alternative for voice channels :)

**Comment by BloodyIron** [comment]
Still a feature desired there @anoadragon453 ;) I just figured I'd stop bugging you about it... for a little bit ;P

I for one am switching away from mumble, and would prefer that PTT not be tied to it, and that any sort of Mumble interfacing be created as another request, as this is originally about PTT.

**Comment by Mathias-g** [comment]
Can you link to the PR? @anoadragon453 

If you don't mind it would be great to chat a bit about it with you so we can help out pushing it across the finishing line.

**Comment by anoadragon453** [comment]
@Mathias-g 

There's two PRs against two separate repositories:

* https://github.com/vector-im/riot-web/pull/10369
* https://github.com/matrix-org/matrix-react-sdk/pull/2280

I believe pretty much everything from the riot-web PR (which mostly dealt with Electron stuff) needs to now be ported to the Riot Desktop repo: https://github.com/vector-im/riot-

However, the biggest hurdle with this PR was getting [iohook](https://github.com/wilix-team/iohook), the native node module used to listen to keystrokes system-wide, to build on Windows. However, https://www.electronjs.org/docs/api/global-shortcut, while insufficient when the PR was initially created, now naively looks like it could suit our uscase. And if so, it makes the implementation much, much simpler.

**Comment by BloodyIron** [comment]
keep at it @anoadragon453 ! :D your work is appreciated!

**Comment by BloodyIron** [comment]
Can we please get this issue tied to a milestone so it doesn't keep getting lost? :(

**Comment by t3chguy** [comment]
That wouldn't mean anything, the team doesn't use github milestones due to them being very limited.

**Comment by Mhowser** [comment]
How about pinning this issue?

**Comment by BloodyIron** [comment]
Would pinning help? I just really want this put in, and the majority of the work is already done, isn't it? :/

**Comment by anoadragon453** [comment]
> the majority of the work is already done, isn't it?

The code has bitrotted so it'll need some freshening up. It was also forked off the riot-web repo, but should not go against element-desktop.

The build process for iohook will also need to be redesigned to align with the native node module stuff that has been paved by the seshat integration.

So it's not a quick job, probably a week or so worth of work, but not impossible. The concepts should still hold.

**Comment by samjco** [comment]
So there is webrtc https://webrtc.github.io/samples/

Or something simple would to just add Pragli 
Here an example of it being using in slack. https://pragli.com/blog/how-to-use-pragli-with-slack/

**Comment by t3chguy** [comment]
Right, Element (Riot) and Jitsi both use WebRTC.

**Comment by BloodyIron** [comment]
Any word on this? :/

**Comment by anoadragon453** [comment]
Not really, just sitting in my backlog unfortunately.

**Comment by BloodyIron** [comment]
So now that it's March 2021, more than 3 years from my original request, can we please get this implemented? The need for it is never going away, it's growing as more people use the ecosystem.

**Comment by anoadragon453** [comment]
I did actually take a look at what would be required to get this working again a couple days ago, and it turns out it shouldn't be too bad.

Watch this space.

**Comment by BloodyIron** [comment]
WOOT!

**Comment by BloodyIron** [comment]
Still watching

**Comment by samjco** [comment]
So this looks as if the implementation is very simple using Google's apprtc :
https://apprtc.appspot.com/params.html


Casing point:
Private/public channels within Riot are also made as voice channels.
Upon user entering, the parameter is pass simply as (random room id: 861000943 ): 
https://appr.tc/r/861000943?audio=true&video=false

My guess is that this could be ran a minimized iframe with the src as the parameter above.
And javascript can be used to manipulate the way the mute button is used to simulate the PPT.
Or 
You can implement your own apprtc within Riot:
https://github.com/webrtc/apprtc

**Comment by BloodyIron** [comment]
@samjco does that PTT method work when focus is away from the application? Say, Element is fully minimised (web/desktop/mobile/wtvr) and someone presses assigned key while another application is focused?

Also, does that method work with the Jitsi back-end currently used? (or even for the P2P 1:1 calling?)

Sorry if this is obvious stuff to you, I'm not a developer, so coming from that perspective.

**Comment by anoadragon453** [comment]
@samjco The work is more based around having Element handle the shortcut and then telling whatever voice/video conferencing application is currently running to mute/unmute. Today the possible options are Jitsi/Matrix 1-1 calls.

Correct me if I'm wrong, but AppRTC - while being quite customisable! - only does 1-1 calls, which doesn't provide too much on top of the existing Matrix 1-1 calls we have today (which do use WebRTC, but use Matrix for signalling).

Both Jitsi and our own stuff have hooks for muting/unmuting programmatically though, so no worries there. Like I said the main work is just getting Element to register an OS-wide shortcut and then providing a notification to the rest of the app that the user has hit the push-to-talk shortcut.

**Comment by BloodyIron** [comment]
@anoadragon453 what do you see as remaining for that then? :)

Also, thanks for the clarification, that is quite enlightening! Sounds like you're still doing well over there in that job :D

**Comment by SebiTimeWaster** [comment]
This is keeping us from using Element.

**Comment by kuipumu** [comment]
Is there any current library or client that is able to do this at the moment?

**Comment by BloodyIron** [comment]
With Element Call coming, will we finally get Push To Talk for VoIP???

**Comment by anoadragon453** [comment]
@BloodyIron Technically, likely not. The hard part of this feature is capturing the press and release of a keyboard shortcut globally even when Element is minimized. Sending the signal up to either a widget or an Element Call class is relatively easy.

However, with calls happening more inside Element as a result of integrating Element Call, this feature may become higher priority for the core team (though PRs always welcome!).

**Comment by BloodyIron** [comment]
But how can applications like Discord achieve that and Element cannot?

Man I have so much on my plate, and I am nowhere near up to speed with writing good code for something like this :/

**Comment by anoadragon453** [comment]
> But how can applications like Discord achieve that and Element cannot?
I'm not sure how Discord does it as it's proprietary software, but this feature is technically possible. See the comment history of this issue and [my old PRs](https://github.com/vector-im/element-web/issues/5993#issuecomment-439190850) for a solution that at one point worked.

**Comment by t3chguy** [comment]
Discord has a much larger team and a lot more native code.

**Comment by MikeBender222** [comment]
PTT functionality is a requirement for me to use a voice chat. I found a workaround solution that lets me use PTT, albeit with some extra steps.

This is a AutoHotKey script that mimics PTT usage. Hopefully this can help others in my position, and perhaps even lead to the implementation of PTT within Element some day.

The downsides:
-Requires AHK, and running an external script.
-This mutes and unmutes your microphone in Windows sound devices. Your microphone isn't just muted in Element, but in all other applications as well.
-Closing the script needs to be done with a hotkey. Otherwise simply closing the script will leave your microphone muted in windows and will need to be unmuted through sound devices settings.

If Element had the functionality, I can see it using AHK as a way of hooking into hotkeys system wide, and Element could have the function to mute/unmute microphone that AHK could trigger in some way.

The script should be saved as .AHK file and run with AutoHotKey (requires installation). Windows only ATM.
https://pastebin.com/raw/n0qfuigN

The script requires figuring out your microphone device ID, which you can do by uncommenting the 2nd line before running the script.

**Comment by sudeaux** [comment]
I'm going to bump this and second comments from @MikeBender222 on this one. Push-to-talk is an absolute requirement - not just for gamers, but professionals and really any voice or conference functionality. For voice and video conferencing push to talk is considered basic feature parity and is expected functionality.

The only existing workarounds involve AHK scripts some of which are / can be falsely detected as a cheat by Anti-Cheat systems.

Unfortunately this is a big miss and major blocker for adoption of Matrix or Element in general. 150+ comments over 4.5 years without a resolution is very disappointing... hoping this is one day so I can take advantage of what otherwise seems like an incredible platform!

**Comment by Heinz14** [comment]
To work around that issue and get global PTT, in Windows you can use AutoHotkey with this script:
[https://superuser.com/questions/1478630/turn-on-mic-only-when-key-is-pressed-like-a-global-push-to-talk-windows-10]

**Comment by JoshuaFern** [comment]
I'm really surprised this hasn't been implemented yet. Every time I look into using Element over Discord this issue has stopped me again and again. Are there any other Matrix clients that have implemented this?

**Comment by kongo09** [comment]
@BloodyIron asked me here https://github.com/vector-im/element-web/issues/2320#issuecomment-1813105031 on my view about PTT. We are aware of the use cases and also have a number of customers potentially interested in it.

However, due to focus on getting Element Call with full Matrix based e2ee implemented, we're deprioritizing other VoIP features like PTT. While Element Call is making great progress as you might have read here: https://element.io/blog/element-x-now-with-embedded-voip/ it is still far from where we would like it to be. Unfortunately, resources are limited and I fully understand the frustration, but right now this is not on the roadmap.

**Comment by BloodyIron** [comment]
> @BloodyIron asked me here [#2320 (comment)](https://github.com/vector-im/element-web/issues/2320#issuecomment-1813105031) on my view about PTT. We are aware of the use cases and also have a number of customers potentially interested in it.
> 
> However, due to focus on getting Element Call with full Matrix based e2ee implemented, we're deprioritizing other VoIP features like PTT. While Element Call is making great progress as you might have read here: https://element.io/blog/element-x-now-with-embedded-voip/ it is still far from where we would like it to be. Unfortunately, resources are limited and I fully understand the frustration, but right now this is not on the roadmap.

Thanks for the insights! Still looking forward to PTT. :)

**Comment by samjco** [comment]
what about using this:
https://github.com/colyseus/demo-push-to-talk

**Comment by BloodyIron** [comment]
The thing is that @anoadragon453 already wrote a good chunk of PTT code for this multiple years ago, and it died on the vine (last I checked, sorry Anoa). So last I checked this is less about the method/library. But I likely have stale insights.

**Comment by samjco** [comment]
Hmm, I was looking at Element X. Nice features, but no demonstration on Voip calls though.
What do you think about integrating 3CX?
https://www.3cx.com/free/

**Comment by iaxat** [comment]
Any update to bring discord like voice channel here?

**Comment by BloodyIron** [comment]
6 years later. I'm pretty sure the devs are ignoring this now, by choice.

**Comment by Thatoo** [comment]
No, it's most probably because they are busy to dev the next voice/video call system called Element Call that will be integrated to Element apps familly... And for what I recall this system does have a push-to-talk system

**Comment by Heinz14** [comment]
> No, it's most probably because they are busy to dev the next voice/video call system called Element Call that will be integrated to Element apps familly... And for what I recall this system does have a push-to-talk system

That's a very bad excuse for refusing to implement a standard feature for 6 years.
Do you happen to have anything to do with that?

**Comment by Thatoo** [comment]
Nope, I'm just trying to figure out what can be priority of Element devs, knowing no humans can do everything at the same time.
I have no idea how long time dev PTT on jitsi widget in Element could take but certainly enough for them to need keeping this time to focus on other things like element call (which, again, is going to replace jitsi).

**Comment by anoadragon453** [comment]
All the VoIP dev is indeed on Element Call. Currently they're working on making it more reliable, see:

- https://github.com/element-hq/synapse/pull/17326 for informing the homeserver that a user has disconnected from the call, instead of users getting stuck.
- https://github.com/element-hq/synapse/pull/17513 to allow users to join a call from multiple devices without race conditions.

Once the product is reliable and stable, they can focus on adding new features.

Element Web still lacks the infrastructure to read shortcuts from outside the application and call functions off the back of them. I did some work towards this in https://github.com/matrix-org/matrix-react-sdk/pull/2280, https://github.com/element-hq/element-web/pull/7709 but it's bit-rotted now.

If someone would like to take that on, that'd be extremely helpful for getting this feature finally landed. Then once Element Call is better integrated into Element Web, it can be hooked up to the shortcut feature.

**Comment by nekohayo** [comment]
FWIW, on Linux [with GNOME 48](https://release.gnome.org/48/developers/) there is now support for global shortcuts on Wayland, see the [portal API documentation](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.GlobalShortcuts.html) for reference.

**Comment by anoadragon453** [comment]
> [electronjs.org/docs/latest/api/global-shortcut](https://www.electronjs.org/docs/latest/api/global-shortcut)

Though as mentioned above, note that this API doesn't have a hook for when a shortcut is released. Meaning you'd only be able to implement Toggle push-to-talk, rather than hold-and-release.

**Comment by gandalftheplad** [comment]
Boosting this as my whole friend group wants to switch to matrix, but this feature is the only thing holding us back.

**Comment by BR5RK** [comment]
boosting this, cannot believe this necessity has been requested for seven years with no implementation. having released voice comms in the first place without global PTT is absolutely mental. this application is completely and utterly useless for voice comms until global PTT is implemented. please allow binding said global PTT to any key as well, no one is ever going to use their fucking spacebar for PTT, unless they have down syndrome.

**Comment by qrenz** [comment]
same boat as above person +1

**Comment by HammyHavoc** [comment]
> > [electronjs.org/docs/latest/api/global-shortcut](https://www.electronjs.org/docs/latest/api/global-shortcut)
> 
> Though as mentioned above, note that this API doesn't have a hook for when a shortcut is released. Meaning you'd only be able to implement Toggle push-to-talk, rather than hold-and-release.

Whilst I'm not a Discord user (I have an account, but no app installed), Discord is also supposedly Electron-based, yet they support PTT via a configurable key-binding, so what gives? https://support.discord.com/hc/en-us/articles/211376518-Voice-Input-Modes-101-Push-to-Talk-Voice-Activated

**Comment by t3chguy** [comment]
Discord isn't open source so we can't see what magic they're doing, likely per-OS native code rather than portable Electron APIs

**Comment by voc0der** [comment]
Hope this doesn't annoy anyone, or violate any rules the following is AI speak, and may be false.

Here's what o3 has to say about this thread:

<html><head></head><body><p><strong>Short answer</strong></p>
<p>Discord doesn’t rely on Electron’s built‑in <code>globalShortcut</code> API at all.<br>
Instead it ships (and statically links) a small native add‑on that installs an OS‑level keyboard hook. The hook can report <em>both</em> the <strong>key‑down</strong> and <strong>key‑up</strong> events for the chosen hot‑key, even while Discord’s window is unfocused, and it passes those events back to the JS layer so the app can start or stop sending audio in real time. Open‑source re‑implementations such as <em>discord‑sandboxed</em> show the same idea in the clear: they use the C / C++‑based <strong>iohook</strong> library to capture global key‑down/‑up and then tell the embedded Discord web‑client to toggle its mic track when the key state changes (<a href="https://github.com/khlam/discord-sandboxed" title="GitHub - khlam/discord-sandboxed: Alternative electron-based Discord client with custom telemetry blocker and privacy-focused push-to-talk.">github.com</a>).</p>
<hr>
<h3>Why the naïve Electron APIs are not enough</h3>

Technique | What you get | What’s missing
-- | -- | --
globalShortcut.register('F24', cb) | Fires once when the key goes down anywhere on the system | No key‑up event, so you can’t know when the user released the PTT key (stackoverflow.com)
webContents.on('before-input-event', …) | Gives both key‑down & key‑up | Only works while the Element window has focus (electronjs.org)
Plain browser JS | Works in focus | Loses all events when the tab isn’t focused – exactly the limitation called out in Element‑web issue #5993 (github.com)


<p>So Element‑web’s authors were technically correct: “Electron doesn’t have a PTT API” – meaning no <strong>built‑in</strong> high‑level API that delivers global key‑down <em>and</em> key‑up out of the box.</p>
<hr>
<h3>What Discord is doing instead</h3>
<ol>
<li>
<p><strong>Native hook</strong><br>
<em>On Windows</em> they inject a DLL that calls <code>SetWindowsHookEx(WH_KEYBOARD_LL)</code>;<br>
<em>on macOS</em> they use a Quartz Event‑Tap;<br>
<em>on X11</em> they grab the key with <code>XGrabKey</code>; each platform handler is built into the app’s “discord_utils” module (the exact symbol names differ per build).</p>
</li>
<li>
<p><strong>Bridge to JS</strong><br>
The hook posts a tiny struct (<code>{keyCode, isDown}</code>) to a Node native module (you can see the same strategy in the open‑source <em>iohook</em> binary that discord‑sandboxed links against).</p>
</li>
<li>
<p><strong>Audio gating</strong><br>
Discord’s voice engine captures audio continuously but drops (encodes silence) until the hook says <em>key‑down</em>. When it receives <em>key‑up</em> it instantly mutes again. Because the engine already runs in native code this change happens with negligible latency.</p>
<blockquote>
<p><em>“When your push‑to‑talk key is held down, the renderer process will send a <code>backspace</code> key‑down … When it’s released the mic is closed again.”</em> (<a href="https://github.com/khlam/discord-sandboxed" title="GitHub - khlam/discord-sandboxed: Alternative electron-based Discord client with custom telemetry blocker and privacy-focused push-to-talk.">github.com</a>)</p>
</blockquote>
</li>
</ol>
<hr>
<h3>How you could replicate this in Element Desktop</h3>
<p>If you only care about PTT <strong>while Element is focused</strong> you can already do it:</p>
<pre><code class="language-ts">// in preload or renderer
window.addEventListener('keydown',  e =&gt; {
  if (e.code === 'F24') localMediaTrack.enabled = true;
});
window.addEventListener('keyup',    e =&gt; {
  if (e.code === 'F24') localMediaTrack.enabled = false;
});
</code></pre>
<p>…but the real demand is global‑scope PTT.<br>
Here’s the typical recipe many Electron apps follow today:</p>
<pre><code class="language-ts">// main.js
import iohook from 'iohook';           // or   node-global-key-listener
iohook.on('keydown',  ev =&gt; {
  if (ev.keycode === 66)  // example: F8
    win.webContents.send('ptt-state', true);
});
iohook.on('keyup',    ev =&gt; {
  if (ev.keycode === 66)
    win.webContents.send('ptt-state', false);
});
iohook.start();

// preload.js
const { contextBridge, ipcRenderer } = require('electron');
contextBridge.exposeInMainWorld('ptt', {
  onState: cb =&gt; ipcRenderer.on('ptt-state', (_, s) =&gt; cb(s))
});

// renderer
window.ptt.onState(state =&gt; {
  localMediaTrack.enabled = state;  // toggle mic instantly
});
</code></pre>
<ul>
<li>
<p><strong>iohook</strong> (or the lighter <strong>node‑global‑key‑listener</strong>) gives you dependable DOWN/UP events across Windows, macOS and X11 Linux (<a href="https://www.npmjs.com/package/%40vizy.io/iohook?utm_source=chatgpt.com" title="@vizy.io/iohook - npm">npmjs.com</a>, <a href="https://www.npmjs.com/package/node-global-key-listener?utm_source=chatgpt.com" title="node-global-key-listener - NPM">npmjs.com</a>).</p>
</li>
<li>
<p>On Wayland, <em>no</em> global hot‑key protocol exists yet, so even Discord falls back to “works only while focused” (users complain about this often) (<a href="https://www.reddit.com/r/linux_gaming/comments/z3jqeu/regarding_discord_on_wayland_wrapperweb/" title="Regarding Discord on Wayland &amp; wrapper/web alternative : r/linux_gaming">reddit.com</a>).</p>
</li>
</ul>
<hr>
<h3>Take‑aways</h3>
<ul>
<li>
<p>Electron itself can only give you <em>pressed‑once</em> accelerators globally or full key‑down/‑up <strong>locally</strong>.</p>
</li>
<li>
<p>To do real hold‑to‑talk you must add a small native component – just like Discord does.</p>
</li>
<li>
<p>Libraries such as <strong>iohook</strong> hide the platform details and are MIT‑licensed, so Element Desktop could ship them without re‑licensing headaches.</p>
</li>
<li>
<p>Under Wayland this still won’t be perfect until a standard “global shortcut” protocol lands upstream, so any solution will have the same focus‑bound limitation Discord has today.</p>
</li>
</ul>
<p>That’s the whole trick: it isn’t magic inside Electron – it’s an extra native hook bolted onto it.</p></body></html>**Short answer**

Discord doesn’t rely on Electron’s built‑in `globalShortcut` API at all.
Instead it ships (and statically links) a small native add‑on that installs an OS‑level keyboard hook. The hook can report *both* the **key‑down** and **key‑up** events for the chosen hot‑key, even while Discord’s window is unfocused, and it passes those events back to the JS layer so the app can start or stop sending audio in real time. Open‑source re‑implementations such as *discord‑sandboxed* show the same idea in the clear: they use the C / C++‑based **iohook** library to capture global key‑down/‑up and then tell the embedded Discord web‑client to toggle its mic track when the key state changes ([[github.com](https://github.com/khlam/discord-sandboxed)][1]).

---

### Why the naïve Electron APIs are not enough

| Technique                                     | What you get                                               | What’s missing                                                                                                               |
| --------------------------------------------- | ---------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------- |
| **`globalShortcut.register('F24', cb)`**      | Fires once when the key *goes down* anywhere on the system | No key‑up event, so you can’t know when the user released the PTT key ([[stackoverflow.com](https://stackoverflow.com/questions/70372032/electron-global-shortcut-alternative-without-reserving-the-key?utm_source=chatgpt.com)][2])                               |
| **`webContents.on('before-input-event', …)`** | Gives both key‑down & key‑up                               | Only works while the Element window has focus ([[electronjs.org](https://electronjs.org/docs/latest/api/web-contents?utm_source=chatgpt.com)][3])                                                          |
| **Plain browser JS**                          | Works in focus                                             | Loses all events when the tab isn’t focused – exactly the limitation called out in Element‑web issue #5993 ([[github.com](https://github.com/element-hq/element-web/issues/5993?utm_source=chatgpt.com)][4]) |

So Element‑web’s authors were technically correct: “Electron doesn’t have a PTT API” – meaning no **built‑in** high‑level API that delivers global key‑down *and* key‑up out of the box.

---

### What Discord is doing instead

1. **Native hook**
   *On Windows* they inject a DLL that calls `SetWindowsHookEx(WH_KEYBOARD_LL)`;
   *on macOS* they use a Quartz Event‑Tap;
   *on X11* they grab the key with `XGrabKey`; each platform handler is built into the app’s “discord\_utils” module (the exact symbol names differ per build).

2. **Bridge to JS**
   The hook posts a tiny struct (`{keyCode, isDown}`) to a Node native module (you can see the same strategy in the open‑source *iohook* binary that discord‑sandboxed links against).

3. **Audio gating**
   Discord’s voice engine captures audio continuously but drops (encodes silence) until the hook says *key‑down*. When it receives *key‑up* it instantly mutes again. Because the engine already runs in native code this change happens with negligible latency.

   > *“When your push‑to‑talk key is held down, the renderer process will send a `backspace` key‑down … When it’s released the mic is closed again.”* ([[github.com](https://github.com/khlam/discord-sandboxed)][1])

---

### How you could replicate this in Element Desktop

If you only care about PTT **while Element is focused** you can already do it:

```ts
// in preload or renderer
window.addEventListener('keydown',  e => {
  if (e.code === 'F24') localMediaTrack.enabled = true;
});
window.addEventListener('keyup',    e => {
  if (e.code === 'F24') localMediaTrack.enabled = false;
});
```

…but the real demand is global‑scope PTT.
Here’s the typical recipe many Electron apps follow today:

```ts
// main.js
import iohook from 'iohook';           // or   node-global-key-listener
iohook.on('keydown',  ev => {
  if (ev.keycode === 66)  // example: F8
    win.webContents.send('ptt-state', true);
});
iohook.on('keyup',    ev => {
  if (ev.keycode === 66)
    win.webContents.send('ptt-state', false);
});
iohook.start();

// preload.js
const { contextBridge, ipcRenderer } = require('electron');
contextBridge.exposeInMainWorld('ptt', {
  onState: cb => ipcRenderer.on('ptt-state', (_, s) => cb(s))
});

// renderer
window.ptt.onState(state => {
  localMediaTrack.enabled = state;  // toggle mic instantly
});
```

* **iohook** (or the lighter **node‑global‑key‑listener**) gives you dependable DOWN/UP events across Windows, macOS and X11 Linux ([[npmjs.com](https://www.npmjs.com/package/%40vizy.io/iohook?utm_source=chatgpt.com)][5], [[npmjs.com](https://www.npmjs.com/package/node-global-key-listener?utm_source=chatgpt.com)][6]).
* On Wayland, *no* global hot‑key protocol exists yet, so even Discord falls back to “works only while focused” (users complain about this often) ([[reddit.com](https://www.reddit.com/r/linux_gaming/comments/z3jqeu/regarding_discord_on_wayland_wrapperweb/)][7]).

---

### Take‑aways

* Electron itself can only give you *pressed‑once* accelerators globally or full key‑down/‑up **locally**.
* To do real hold‑to‑talk you must add a small native component – just like Discord does.
* Libraries such as **iohook** hide the platform details and are MIT‑licensed, so Element Desktop could ship them without re‑licensing headaches.
* Under Wayland this still won’t be perfect until a standard “global shortcut” protocol lands upstream, so any solution will have the same focus‑bound limitation Discord has today.

That’s the whole trick: it isn’t magic inside Electron – it’s an extra native hook bolted onto it.

[1]: https://github.com/khlam/discord-sandboxed "GitHub - khlam/discord-sandboxed: Alternative electron-based Discord client with custom telemetry blocker and privacy-focused push-to-talk."
[2]: https://stackoverflow.com/questions/70372032/electron-global-shortcut-alternative-without-reserving-the-key?utm_source=chatgpt.com "Electron Global Shortcut alternative without reserving the key"
[3]: https://electronjs.org/docs/latest/api/web-contents?utm_source=chatgpt.com "webContents | Electron"
[4]: https://github.com/element-hq/element-web/issues/5993?utm_source=chatgpt.com "Push-to-talk in video/voice conferencing · Issue #5993 - GitHub"
[5]: https://www.npmjs.com/package/%40vizy.io/iohook?utm_source=chatgpt.com "@vizy.io/iohook - npm"
[6]: https://www.npmjs.com/package/node-global-key-listener?utm_source=chatgpt.com "node-global-key-listener - NPM"
[7]: https://www.reddit.com/r/linux_gaming/comments/z3jqeu/regarding_discord_on_wayland_wrapperweb/ "Regarding Discord on Wayland & wrapper/web alternative : r/linux_gaming"

**Comment by t3chguy** [comment]
There was a prior attempt to use iohook https://github.com/element-hq/element-web/pull/10369 - not sure what went wrong with it

**Comment by anoadragon453** [comment]
It was primarily due to being a pain to get working (i.e. building) on Windows. I'm sure things are different/easier 6 years on and recommend someone give it another shot 🙂

**Comment by HammyHavoc** [comment]
> It was primarily due to being a pain to get working (i.e. building) on Windows. I'm sure things are different/easier 6 years on and recommend someone give it another shot 🙂

I'm sure it will still present challenges, but if someone does manage to pull this off then it'll be an absolutely huge win for both Matrix and Element adoption in terms of gamers and enterprise demographics. It's really the only thing that keeps a lot of people I know firmly entrenched on Discord as they're using it in-game for voice chat with their friends and community.

**Comment by Dark-Sheepa** [comment]
Honestly. This is a feature I've been following for several years, and it's also the thing that will determine when most of my friends switch from Discord to Matrix. I'm very curious to see if I'll ever live to see it, because it still has no end in sight, and who knows if it ever will.. 😩

**Comment by BloodyIron** [comment]
Yeah what's going on here? It's now more than 7 years since the request was made and clearly it's popular enough to warrant implementation... so...???

**Comment by Half-Shot** [comment]
Last time I checked there still isn't a competent cross-platform library to do global keyboard hooks. Discord has invested efforts to build one for themselves, but decided not to share it with the rest of the world. I've been checking in every so often to see if one emerges but Element presently doesn't have the time to build and maintain one, and we can't achieve push-to-talk without it.

The march of time alone does not bring new features; active participation does.

**Comment by k3d3** [comment]
> Last time I checked there still isn't a competent cross-platform library to do global keyboard hooks.

Sure, but we could start somewhere. If there was just a windows-only, or just a Linux-only PTT implementation, that alone will entice more people to contribute the missing pieces than to have nothing and expect that we'd need them all at once to even begin thinking about it.

Even the baseline implementation (assuming no cross platform global shortcuts), I'm certain people are willing to contribute implementations, but there needs to be a design or guideline for push-to-talk. As far as I know, no such design exists, and no contributor will want to make a large implementation from scratch if there's a risk it'll be rejected outright.

**Comment by Dark-Sheepa** [comment]
As a game developer, I have a huge community of people around me, and whenever we discuss the possibility of leaving Discord, we always come up against voice communication, which has already been addressed in past months, but without the necessary features that people are unfortunately already used to ... don't want to give up. I think that if Element needs a new batch of people who are willing to pay them for a premium, which many users I know on Discord, including myself, do, someone from the team should think about it. Otherwise, we can wait another 8 years and only target companies. 🤷‍♂️  :))

I would be happy to help with some..., but I'm more of a 3D graphic designer and animator than a programmer.

**Comment by BloodyIron** [comment]
Something > 0.

**Comment by Eddudos** [comment]
Bringing this up once again.
Correct me if I'm wrong, but I feel like such functionality also could be achieved writing some simple browser extension. 
It could be really cool since you can use it on any OS via chrome-based browser. 

Especially, since implementing it natively in browser (or electron based) Jitsi requires architecture planning, writing different native listeners and all that complex stuff. I think I will try to research if it's possible

**Comment by Eddudos** [comment]
Okay, I've forked the similar repo, but for google meet and simply adapted it to the Jitsi and it worked! 
Of course, that shouldn't be considered as a proper way of solving this issue, but since there are no alternatives, it might help some people just to enjoy Jitsi by toggle muting globally!

Tested on MacOS Sonoma 14.5 in Brave browser. Works for Jitsi in element and also in Jitsi web. You can also add selfhosted urls in background.js if you need, works for me.

Here is the repo:
https://github.com/Eddudos/global_mute_for_jitsi_meet

P.S. I'm no expert in js and extensions, so repo is open for any PRs and comments. Also, if somebody can publish it in the chrome store it would be great!

**Comment by dannersy** [comment]
I am sure there are organizational reasons why this issue has not been addressed yet, but I cannot help but wonder if Element is missing the forest for the trees when focusing efforts elsewhere.

I would assume Element would want more people to use their software and I have to say, the lack of push-to-talk seems to be a very substantial deal breaker for those looking to move away from Discord as they look to IPO and with their age verification policy. I wish I could say this is just anecdotal, but the new influx of threads about moving away from Discord almost always cite that Element does not support PTT.

I would like to contribute but this is a tad out of my wheelhouse. So until I have the time and brain space, is there anything else that can be done to push this as a priority?

**Comment by t3chguy** [comment]
Contributions welcome. Our efforts are prioritised on our target markets which currently does not include being a Discord replacement, doesn't mean it can't be one, just unlikely to see it on any roadmaps soon given focus for this calendar year is elsewhere

**Comment by Heinz14** [comment]
Answer from Element after just 2 minutes. Yes, the priorities are set completely wrong.

**Comment by t3chguy** [comment]
You really expect to micromanage my time when I get an email notification, especially when on the triage rota, and I'm waiting for CI to build? You are the problem with open source.

**Comment by skylord123** [comment]
> Our efforts are prioritised on our target markets which currently does not include being a Discord replacement

That’s reasonable, and I understand that not everything can be prioritized at once.

That said, with Discord rolling out ID verification in more regions, there’s clearly a wave of users looking for alternatives right now. Even if “Discord replacement” isn’t the goal, features like push-to-talk are table stakes for a lot of gaming and community servers. There’s a real opportunity for Matrix/Element to pick up some of that momentum.

I do wish there were a more visible bounty or funding mechanism for long-standing feature requests like this. I’d happily contribute financially toward getting push-to-talk implemented, and I suspect others would too. If there were an official or semi-official way to pool funds for specific issues, it might help bridge the gap between community demand and roadmap priorities.

**Comment by HammyHavoc** [comment]
> > Our efforts are prioritised on our target markets which currently does not include being a Discord replacement
> 
> That’s reasonable, and I understand that not everything can be prioritized at once.
> 
> That said, with Discord rolling out ID verification in more regions, there’s clearly a wave of users looking for alternatives right now. Even if “Discord replacement” isn’t the goal, features like push-to-talk are table stakes for a lot of gaming and community servers. There’s a real opportunity for Matrix/Element to pick up some of that momentum.
> 
> I do wish there were a more visible bounty or funding mechanism for long-standing feature requests like this. I’d happily contribute financially toward getting push-to-talk implemented, and I suspect others would too. If there were an official or semi-official way to pool funds for specific issues, it might help bridge the gap between community demand and roadmap priorities.
> 

I know the sentiment of one rather paying to fund FOSS dev than pay for Discord Nitro isn't an uncommon one these days.

**Comment by Heinz14** [comment]
> Our efforts are prioritised on our target markets which currently does not include being a Discord replacement

Which "target market" doesn't need the standard feature push-to-talk?

**Comment by turt2live** [comment]
I believe the product team is well aware of the desire for this feature, so I'm locking this to avoid further comments which drive the conversation in circles.
